contractcourt/chain_watcher.go

```diff
@@ -287,17 +287,32 @@ func (c *chainWatcher) Start() error {
 		}
 	}
 
-	localKey := chanState.LocalChanCfg.MultiSigKey.PubKey.SerializeCompressed()
-	remoteKey := chanState.RemoteChanCfg.MultiSigKey.PubKey.SerializeCompressed()
-	multiSigScript, err := input.GenMultiSigScript(
-		localKey, remoteKey,
+	localKey := chanState.LocalChanCfg.MultiSigKey.PubKey
+	remoteKey := chanState.RemoteChanCfg.MultiSigKey.PubKey
+
+	var (
+		pkScript []byte
+		err      error
 	)
-	if err != nil {
-		return err
-	}
-	pkScript, err := input.WitnessScriptHash(multiSigScript)
-	if err != nil {
-		return err
+	if chanState.ChanType.IsTaproot() {
+		pkScript, _, err = input.GenTaprootFundingScript(
+			localKey, remoteKey, 0,
+		)
+		if err != nil {
+			return err
+		}
+	} else {
+		multiSigScript, err := input.GenMultiSigScript(
+			localKey.SerializeCompressed(),
+			remoteKey.SerializeCompressed(),
+		)
+		if err != nil {
+			return err
+		}
+		pkScript, err = input.WitnessScriptHash(multiSigScript)
+		if err != nil {
+			return err
+		}
 	}
 
 	spendNtfn, err := c.cfg.notifier.RegisterSpendNtfn(
```

Comment:
For taproot pkscripts, `RegisterSpendNtfn` will only notify on the outpoint being spent and not the outpoint+pkscript due to `ComputePkScript` being unable to compute the pkscript if a keyspend is used. We need to add a `RegisterConfirmationsNtfn` here to ensure that the outpoint+pkscript pair is confirmed before calling `RegisterSpendNtfn`. Otherwise, it's possible an attacker gives us an outpoint in `funding_created` that doesn't actually spend to the pkscript and then spends this outpoint with unexpected witness data since:
 - `RegisterSpendNtfn` doesn't actually verify the pkscript has been spent from, and
 - `chain_watcher` starts watching the pkscript before the outpoint+pkscript is verified as confirmed in the funding manager (and we can't change this in case there's a quick channel open + close)
 
This is similar to an old DoS in https://github.com/lightningdevkit/rust-lightning/pull/1585 and I think we should just eliminate the possibility of an attacker doing this even if there's not a clear-cut DoS in lnd.

Reply:
> For taproot pkscripts, RegisterSpendNtfn will only notify on the outpoint being spent and not the outpoint+pkscript due to ComputePkScript being unable to compute the pkscript if a keyspend is used.

FWIW that only applies for neutrino nodes. Also recall that we have the `chanvalidate` logic which handles that for funding transaction confirmation. In this case, given we know the outpoint which itself was something that we explicitly signed off on (we signed the commitment transaction which is the effective transaction root for everything), I don't see how this applies. 

> Otherwise, it's possible an attacker gives us an outpoint in funding_created that doesn't actually spend to the pkscript and then spends this outpoint with unexpected witness data since:

I don't see how this does anything for them: `chanvalidate` [will catch this](https://github.com/lightningnetwork/lnd/blob/master/funding/manager.go#L2843-L2850), and then never send the channel to the link, nor mark it as being confirmed.

Reply:
In this case `chanvalidate` won't run since the `chain_watcher` watches for a spend of the outpoint regardless of whether the funding manager validated it (it's started in `WatchNewChannel` [here](https://github.com/lightningnetwork/lnd/blob/2bb28f7aaff7388137835a9ca724d8906b27fc93/funding/manager.go#L2312)). This applies for _all_ nodes because the `txnotifier` logic uses the [zero taproot pkscript](https://github.com/lightningnetwork/lnd/blob/2bb28f7aaff7388137835a9ca724d8906b27fc93/chainntnfs/txnotifier.go#L341). This means they can give a random outpoint in `funding_created` and then spend from it with a tx that we don't expect. Depending on the spending conditions here, the `chain_watcher` will think it's a [coop close, an unknown local state, or an unknown remote state](https://github.com/lightningnetwork/lnd/blob/a6fc5d0753d6c28416725e9c5de29d15c7cbca0c/contractcourt/chain_watcher.go#L641-L682). It will then send a signal to the `channel_arbitrator` to act on one of these events.

> I don't see how this does anything for them

I don't immediately see a way that this is a DoS, but this is more about sandboxing to ensure a DoS can't ever happen by making the `chain_watcher` act on a channel iff it spends to the correct pk script. For example, if we had a change in the future that:
- set a field when the channel is marked as confirmed in the funding manager
- use this field in the `channel_arbitrator`,

we would panic if a non-marked-as-confirmed channel made it to the channel arbitrator. By registering for a confirmation ntfn, we ensure that the outpoint has the correct pk script and that an unexpected spending tx can't be used and send fake notifications to the channel arbitrator. The attack I'm describing is _basically_ the same as the one described in the LDK link.